lib/phase/merge.js

```diff
@@ -70,6 +71,82 @@ function isEmpty(value) {
 function mergePipelineLevelConfig(parsedDoc, pipelineTemplate) {
     const newPipeline = pipelineTemplate.config;
 
+    // Handle jobs config
+    // Check if job names are valid
+    if (parsedDoc.jobs) {
+        const pipelineTemplateJobNames = Object.keys(newPipeline.jobs);
+        const invalidJobNames = [];
+
+        Object.keys(parsedDoc.jobs).forEach(jobName => {
+            if (!pipelineTemplateJobNames.includes(jobName)) {
+                invalidJobNames.push(jobName);
+            }
+        });
+
+        if (invalidJobNames.length > 0) {
+            throw new Error(
+                `User template has job name(s) that do not exist in pipeline template ${parsedDoc.template}: ${invalidJobNames}`
+            );
+        }
+
+        // Flatten jobs
+        const fieldsToFlatten = ['image', 'requires'];
+
+        Object.keys(parsedDoc.jobs).forEach(jobName => {
+            const newJob = clone(pipelineTemplate.config.jobs[jobName]);
+            const oldJob = clone(parsedDoc.jobs[jobName]);
+
+            // Replace
+            fieldsToFlatten.forEach(key => {
+                if (oldJob[key]) {
+                    newJob[key] = oldJob[key];
+                }
+            });
+
+            // Merge job environment
+            if (oldJob.environment || newJob.environment) {
+                newJob.environment = {
+                    ...newJob.environment,
+                    ...oldJob.environment
+                };
+            }
+
+            // Replace settings
+            const yamlSettingsConfig = oldJob.settings || {};
+            const templateSettingsConfig = newJob.settings || {};
+            const newSlackSettings = yamlSettingsConfig.slack || {};
+            const oldSlackSettings = templateSettingsConfig.slack || {};
+            const newEmailSettings = yamlSettingsConfig.email || {};
+            const oldEmailSettings = templateSettingsConfig.email || {};
+
+            newJob.settings = {
+                ...templateSettingsConfig,
+                ...yamlSettingsConfig
+            };
+
+            if (!isEmpty(oldSlackSettings)) {
```

Comment:
Do we need the not empty check, `oldSlackSettings` and `oldEmailSettings` are already set to {}, so expanding them `...{}` should be fine

Reply:
I think it's still needed here. Otherwise we get an error `\"slack.0\" is not allowed`